katdal/visdatav4.py

```diff
@@ -76,6 +76,57 @@ def _calc_azel(cache, name, ant):
     return sensor_data
 
 
+def _calc_delay(cache, name, inp):
+    """Extract virtual applied delay/phase sensors from raw CBF sensors."""
+    # Obtain the relevant CBF attributes from the cache
+    instrument = cache.get('Correlator/instrument')[0]
+    sync_time = cache.get('Correlator/sync_time')[0]
+    scale_factor_timestamp = cache.get('Correlator/scale_factor_timestamp')[0]
+    # Don't extract the sensor since the real timestamps are hidden in the values
+    getter = cache.get(f'{instrument}_antenna_channelised_voltage_{inp}_delay',
+                       extract=False)
+    sensor_data = getter.get()
+    # Unwrap and unpack the five components of the CBF sensor
+    values = [ComparableArrayWrapper.unwrap(v) for v in sensor_data.value]
+    adc_sample_counts, delays, delay_rates, phases, phase_rates = zip(*values)
+    # Convert the ADC sample count of each delay update to proper Unix timestamp
+    times = sync_time + np.array(adc_sample_counts) / scale_factor_timestamp
+    # Use actual delay/phase rate used by F-engine to interpolate between updates.
+    # Insert interpolation endpoint just before next update for strict monotonicity.
+    next_times = np.r_[times[1:], 2 * times[-1] - times[-2]] - 1e-6
```

Comment:
This will raise IndexError if there are not at least two samples.

Reply:
Rather than extrapolating forward by an estimating update interval, should the last rate not be extrapolated forward to the final dump? I'm thinking about a case where something goes wrong and delay updates stop happening and the last delay rate gets applied indefinitely.

Reply:
- I wilfully ignored the first issue, but I'll try to patch it up.
- The second suggestion is great, since that is exactly what will happen. Will do!